cli/src/build_state_records.rs

```diff
@@ -0,0 +1,336 @@
+//! Module to manage the track changes between each run by comparing the files' states
+//! and the included additional features between the current and the previous build run.
+//! File comparison is done by calculating the checksum of the files on each target then comparing
+//! it with persisted checksum from the last run.  
+//!
+//! This module manages saving the loading the state records as well.
+
+use std::{
+    collections::{btree_map, BTreeMap, BTreeSet},
+    fs::File,
+    io::{BufReader, BufWriter},
+    ops::Deref,
+    path::PathBuf,
+    sync::{Mutex, OnceLock},
+};
+
+use anyhow::{anyhow, Context};
+use console::style;
+use dir_checksum::calc_combined_checksum;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    job_type::JobType, jobs_runner::additional_features::AdditionalFeatures, location::get_root,
+    target::Target, JobsState,
+};
+
+/// Deprecated filenames that were used previously.
+const DEPRECATED_FILE_NAMES: [&str; 2] = [".build_chksum_dev", ".build_chksum_prod"];
+
+/// Name of the file used to save the state of the last build run.
+const PERSIST_FILE_NAME: &str = ".build_last_state";
+
+#[derive(Debug, Serialize, Deserialize)]
+/// Manages and compares the file states for the targets between current and previous builds.
+/// It calculates the checksums of the files for each targets and saves them to a file after
+/// each build, and for the next build it'll calculate the checksum again and compare it with
+/// the saved one.
+/// It also manages loading and clearing the saved checksum records as well.
+pub struct BuildStateRecords {
+    items: Mutex<BuildStateItems>,
+}
```

Comment:
Bellow in code you are getting access to `BuildStateRecords` as a ref

```
 pub fn get(production: bool) -> anyhow::Result<&'static BuildStateRecords> {
        static CHECKSUM_RECORDS: OnceLock<anyhow::Result<BuildStateRecords>> = OnceLock::new();

        CHECKSUM_RECORDS
            .get_or_init(|| BuildStateRecords::load(production))
            .as_ref()
            .map_err(|err| anyhow!("{err}"))
    }
```

All looks good, but it's still a little bit weird. You are wrapping `items` into `Mutex` to have `mut` access to a list, all good. But what it in some time letter you will need some new member in `BuildStateRecords`, which also would need `mut` access? 

Would it make the solution more scalable, if in `OnceLock` you will create not an instance of `BuildStateRecords`, but `Mutex<BuildStateRecords>`. In this case, you would not need to wrap `items` into `Mutex` and you will be able to get access to `&mut BuildStateRecords` from everywhere. Besides that, you `BuildStateRecords` can be easily extended with a new field, if it will be needed. Also good alternative might be `RwLock` if not all calls of `BuildStateRecords` require `mut`.

Reply:
Very thanks @DmitryAstafyev for pointing out the issue with internal Mutex. At the start I thought that we could have other fields which doesn't need to be wrapped within the Mutex, but it ended up with more complex code. I moved the Mutex to outside of the struct itself.
This change won't remove the `as_ref()` call because this call is to avoid wrapping the Mutex within an `Arc` since we always need a reference of the records.

Reply:
@AmmarAbouZor thanks for the corrections. Let it go! You are able to merge it as well